usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -3069,97 +3080,117 @@ arc_evict_state(arc_state_t *state, uint64_t spa, int64_t bytes,
 	num_sublists = multilist_get_num_sublists(ml);
 
 	/*
-	 * If we've tried to evict from each sublist, made some
-	 * progress, but still have not hit the target number of bytes
-	 * to evict, we want to keep trying. The markers allow us to
-	 * pick up where we left off for each individual sublist, rather
-	 * than starting from the tail each time.
+	 * Do the job for each type of headers
+	 * L2_ONLY is used to indicate that we first only
+	 * evict headers which have L2 headers.
+	 * L2_ONLY_NOLIMIT does the same but tries to find more
+	 * such headers ignoring zfs_arc_evict_batch_limit.
+	 * L1_AND_L2 catches both L1 & L2 headers.
 	 */
-	markers = kmem_zalloc(sizeof (*markers) * num_sublists, KM_SLEEP);
-	for (int i = 0; i < num_sublists; i++) {
-		markers[i] = kmem_cache_alloc(hdr_full_cache, KM_SLEEP);
+	for (evict_header_type_t header_type = L2_ONLY;
+	     header_type <= L1_AND_L2 && (total_evicted < bytes || bytes == ARC_EVICT_ALL);
+	     header_type++) {
+		if ((header_type == L2_ONLY || header_type == L2_ONLY_NOLIMIT)
+		    && !arc_evict_l2_first && !arc_evict_l2_only)
+			continue;
+		if (header_type == L1_AND_L2
+		    && arc_evict_l2_only && total_evicted > 0 && bytes != ARC_EVICT_ALL)
```

Comment:
Just to be clear:

```
arc_evict_l2_first switches to the default (current) behaviour if not enough buffers (total_evicted < bytes) have been evicted, so no issue here.
```

is AFTER all the L2-backed buffers for the non-faulted pool have already been evicted, right?  *All* of them?